packages/cli/src/shell.ts

```diff
@@ -0,0 +1,90 @@
+import { DeviceProvider, ScryptedStatic, StreamService } from "@scrypted/types";
+import { createAsyncQueue } from '../../../common/src/async-queue';
+
+export async function connectShell(sdk: ScryptedStatic) {
+    const termSvc = await sdk.systemManager.getDeviceByName<DeviceProvider>("@scrypted/core").getDevice("terminalservice");
+    if (!termSvc) {
+        throw Error("@scrypted/core does not provide a Terminal Service");
+    }
+
+    const termSvcDirect = await sdk.connectRPCObject<StreamService>(termSvc);
+    const dataQueue = createAsyncQueue<Buffer>();
+    const ctrlQueue = createAsyncQueue<any>();
+
+    if (process.stdin.isTTY) {
+        process.stdin.setRawMode(true);
+    } else {
+        process.stdin.on("end", () => {
+            ctrlQueue.enqueue({ eof: true });
+            dataQueue.enqueue(Buffer.alloc(0));
+        });
+    }
+    ctrlQueue.enqueue({ interactive: Boolean(process.stdin.isTTY) });
+
+    const dim = { cols: process.stdout.columns, rows: process.stdout.rows };
+    ctrlQueue.enqueue({ dim });
+
+    let bufferedLength = 0;
+    const MAX_BUFFERED_LENGTH = 64000;
+    process.stdin.on('data', async data => {
+        bufferedLength += data.length;
+        const promise = dataQueue.enqueue(data).then(() => bufferedLength -= data.length);
+        if (bufferedLength >= MAX_BUFFERED_LENGTH) {
+            process.stdin.pause();
+            await promise;
+            if (bufferedLength < MAX_BUFFERED_LENGTH)
+                process.stdin.resume();
+        }
+    });
+
+    async function* generator() {
+        while (true) {
+            const ctrlBuffers = ctrlQueue.clear();
+            if (ctrlBuffers.length) {
+                for (const ctrl of ctrlBuffers) {
+                    if (ctrl.eof) {
+                        // flush the buffer before sending eof
+                        const dataBuffers = dataQueue.clear();
```

Comment:
Oh, I hadn't considered the message ordering mattering between data and control. was this added for non-pty pipe handling?

Reply:
Yeah since the ctrl messages take priority, we should check that any data in the buffer should be flushed for EOF case, since otherwise it will not get processed in the remote after the EOF message